src/shared/pcre2-util.c

```diff
@@ -33,12 +35,118 @@ int dlopen_pcre2(void) {
                         DLSYM_ARG(pcre2_get_error_message),
                         DLSYM_ARG(pcre2_match),
                         DLSYM_ARG(pcre2_get_ovector_pointer));
+#else
+        return -EOPNOTSUPP;
+#endif
 }
 
+int pattern_compile_and_log(const char *pattern, PatternCompileCase case_, pcre2_code **ret) {
+#if HAVE_PCRE2
+        PCRE2_SIZE erroroffset;
+        pcre2_code *p;
+        unsigned flags = 0;
+        int errorcode, r;
+
+        assert(pcre2_dl);
+        assert(pattern);
+
+        if (case_ == PATTERN_COMPILE_CASE_INSENSITIVE)
+                flags = PCRE2_CASELESS;
+        else if (case_ == PATTERN_COMPILE_CASE_AUTO) {
+                _cleanup_(sym_pcre2_match_data_freep) pcre2_match_data *md = NULL;
+                bool has_case;
+                _cleanup_(sym_pcre2_code_freep) pcre2_code *cs = NULL;
+
+                md = sym_pcre2_match_data_create(1, NULL);
+                if (!md)
+                        return log_oom();
+
+                r = pattern_compile_and_log("[[:upper:]]", PATTERN_COMPILE_CASE_SENSITIVE, &cs);
+                if (r < 0)
+                        return r;
+
+                r = sym_pcre2_match(cs, (PCRE2_SPTR8) pattern, PCRE2_ZERO_TERMINATED, 0, 0, md, NULL);
+                has_case = r >= 0;
+
+                flags = !has_case * PCRE2_CASELESS;
```

Comment:
This part should probably be extracted and the compiled pattern cached.

Reply:
pattern_compile_and_log() shouldn't be on the hot path, since we should only compile patterns once. If we'd need to cached this pattern, we'd either need to bother every user with taking care of freeing the pattern etc, or use thread local cache variables, at which point valgrind might start to complain because the value is never freed (which we could fix by calling a free function at the end of the process, but then we'd need to take care to do so everywhere, which would be a burden as well.

So if this was a hot path we should definitely do this, but since it isn't, I figure it's not a problem to compile this pattern every time we need to compile another pattern. Also, this is only done if the "smart" logic is used, which we'll probably only do in journalctl, and not in the journald log regex stuff I think?